src/algorithms/path_contexts.py

```diff
@@ -0,0 +1,331 @@
+    :param token_extractor: function to transform a node into a single string token
+    :param leaf_token: get leaves token as a different node
+    :return: list(tuple) list of paths context like (u, path, v) where u & v and the starting and
+     ending leaf, and path is the list of nodes in their minimal distance path.
+    """
+
+    tree, leaves = extend_tree(uast)
+
+    paths = []
+    if len(leaves) > 1:
+        for i in range(len(leaves)):
+            for j in range(i + 1, min(i + max_width, len(leaves))):
+                u, v = leaves[i], leaves[j]
+                # TODO decide where to filter comments and decouple bblfsh
+                if not is_noop_line(u) and not is_noop_line(v):
+                    ancestor = lca(u, v)
+                    d = distance(u, v, ancestor)
+                    if d <= max_length:
+                        u, path, v = get_path(u, v, ancestor, leaf_token=leaf_token)
+                        # convert nodes to its desired representation
+                        paths.append((u, [token_extractor(p) for p in path], v))
+
+    return paths
+
+
+# aux function to print paths
+def print_paths(paths, symbols_repr={"UP": "↑", "DOWN": "↓"}):
+    for u, path, v in paths:
+        print(u, end="")
+        for element in path:
+            if type(element) == str and element in symbols_repr.keys():
+                print("%s " % symbols_repr[element], end="")
+            else:
+                print("[%s] " % element, end="")
+
+        print(v)
+
+
+def get_unique_tokens_and_paths(fp):
+    _, paths = fp
+
+    uniq_tokens = set()
+    uniq_paths = set()
+    for u, path, v in paths:
+        uniq_tokens.add(u)
+        uniq_tokens.add(v)
+        uniq_paths.add(path)
+
+    return uniq_tokens, uniq_paths
+
+
+def write_to_disk(output, file_name, paths):
+    # if file_name has relative paths remove them
+    file_name = file_name.replace("../", "")
+    file_out = os.path.join(output, file_name)
+
+    os.makedirs(os.path.dirname(file_out), exist_ok=True)
+
+    print("Writing id paths to %s" % file_out)
+
+    with open(file_out, 'w') as f:
+        for p in paths:
+            f.write(" ".join([str(e) for e in p]) + "\n")
+    f.close()
+
+
+def extract_paths(data, extension, max_length, max_width):
+    """
+    Computes the code paths of each file and the lists of all tokens seen
+    :param data: folder where the AST files are (for now bblfsh binary UASTs)
+    :param extension: extension to filter by; default uast.bin as saved by 'repos_to_uast.py'
+    :param max_length: max length of the code paths
+    :param max_width: max number of leaves between origin and end of a given path
+    :return: id_paths: tuples with file names and its list of path contexts , token2id: all
+    tokens and their id
+    """
+    files = recursive_glob(data, '*%s' % extension)
+
+    file_paths = [None] * len(files)
+    for i, file in enumerate(files):
+        uast = Node.FromString(open(file, 'rb').read())
```

Comment:
We do not need this code if we keep only one command line entry point. but it case we keep it, it is a not python way to do things:
```python
file_codepaths = dict()
for file in files:
    uast = Node.FromString(open(file, 'rb').read())
    file_codepaths[file] = get_paths(uast, max_length=max_length, max_width=max_width)
```

Reply:
there are many statements similar because I used another framework instead of spark and this is the structure requirements. I'll change them all, just wanted you to check the overall organization (which I copied from sourced.ml without much thought).